modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointPagesWriter.java

```diff
@@ -253,38 +251,21 @@ private void updatePartitions() throws IgniteInternalCheckedException {
         while (!shutdownNow.getAsBoolean() && updatePartitionIds.next(queueResult)) {
             GroupPartitionId partitionId = queueResult.getValue();
 
-            PartitionFilePageStore partitionFilePageStore = partitionFilePageStoreManager.getStore(
-                    partitionId.getGroupId(),
-                    partitionId.getPartitionId()
-            );
-
             ByteBuffer buffer = threadBuf.get();
 
             PersistentPageMemory pageMemory = queueResult.getKey();
 
-            long partitionMetaPageId = pageMemory.partitionMetaPageId(partitionId.getGroupId(), partitionId.getPartitionId());
-
-            long pageAddr = GridUnsafe.bufferAddress(buffer);
-
-            // TODO: IGNITE-16657 Merger will put the page into a partition file.
-            PartitionMetaIo io = PartitionMetaIo.VERSIONS.latest();
+            PartitionMetaSnapshot partitionMetaSnapshot = partitionMetaManager.getMeta(partitionId).metaSnapshot(checkpointProgress.id());
```

Comment:
Ok, you're not gonna like what I want to say. But, a collection of dirty partition ids is not required anymore once you implemented "async" meta snapshots.
Here's what you can do: for each page you need to write you additionally check if its partition has its meta saved. This can be done with a simple `if (saveMetaPage.put(groupPartitionId))`. Then, if current thread is the one that needs to write meta, you just do it by acquiring meta object the same way as it is in the code right now.
The point is, a regular list of dirty pages already contains all information about dirty partitions.
Please ask questions if I didn't manage to explain myself here.

Reply:
Yes, I understand you, it can only create contention in some situations and use more memory due to the arrangement of the **ConcurrentHashSet**, but in general it is reasonable.

Reply:
Well, you already have a collection of ids. Maybe there's an argument that we need to create GroupPartitionId objects all the time, but even that is not a problem I think. Can be easily optimized though. And contention - I don't understand what you mean here specifically. Like during a "put" into a set?

Reply:
Page id consist of more than just grpId and partId.
So objects are problems or not? Above you write that this is not a problem.
Of course put into a set, since not one thread writes to one partition, but many threads write to a partition

Reply:
Fix it